cdisutils/storage.py

```diff
@@ -8,23 +8,130 @@
 TODO: add tests
 
 """
+import hashlib
+import json
+import re
+import socket
+import ssl
 
-from .log import get_logger
-from boto import connect_s3
+from boto.compat import http_client
+from boto import config, UserAgent, https_connection
+from boto.s3.connection import S3Connection
 from boto.s3 import connection
 from dateutil import parser
 from datetime import timedelta, datetime
 from future.utils import iteritems
 from future.standard_library import install_aliases
+
+import cdisutils.log
+import cdisutils.parsers
+
 install_aliases()
 
-from urllib.parse import urlparse
 
-import hashlib
-import json
-import re
+class S3ConnectionProxyFix(S3Connection):
+    """Custom S3Connection class to workaround boto proxy issue.
+
+    This fix will apply to any Python 3 boto code that attempts to interact with
+    Amazon S3 because it needs a proxy to access it.
+
+    With the Python 3 upgrade came breaking changes specifically pertaining to
+    bytes and strings. The sendall() socket command no longer accepts strings.
+    We have to convert them all to bytes before sending.
+
+    There have been attempts to merge a fix for this into the main boto branch
+    but they have not been successful.
+    - https://github.com/boto/boto/pull/2718
+    - https://github.com/boto/boto/pull/3699
+    """
+    def __init__(self, *args, **kwargs):
+        super(S3ConnectionProxyFix, self).__init__(*args, **kwargs)
+        self.logger = cdisutils.log.get_logger("S3ConnectionProxyFix")
+
+    def _get_host(self, host=None, port=None):
+        if host and port:
+            return '%s:%d' % (host, port)
+        else:
+            return '%s:%d' % (self.host, self.port)
```

Comment:
Just a minor style suggestion to use double-quotes here and in a couple other places in this class.